src/devtools/server.ts

```diff
@@ -15,8 +15,53 @@ devtoolsEvents.addListener("fromDevTools", (msg) => {
   Debug.info("DevTools > WS: " + JSON.stringify(msg));
 });
 
-export function startServer(port: number): Disposable {
+let id = 1;
+
+export function sendToDevTools(message: unknown) {
+  devtoolsEvents.emit("toDevTools", {
+    id: `vscode-${id++}`,
+    source: "apollo-client-devtools",
+    type: "actor",
+    message,
+  });
+}
+
+export let serverState:
+  | { port: false | number; disposable: Disposable }
+  | undefined = undefined;
+
+export function startServer(port: number) {
+  const state = {
+    port: false as false | number,
+    disposable: new Disposable(() => {
+      if (wss) {
+        wss.close();
+        wss = null;
+      }
+      if (serverState === state) {
+        serverState = undefined;
+      }
+      sendToDevTools({ type: "port.changed", port, listening: false });
+    }),
+  };
+  if (serverState) {
+    if (serverState.port === port) {
+      // nothing to do
+      return;
+    } else {
+      // changing port, stop the old server
+      serverState.disposable.dispose();
+    }
+  }
```

Comment:
```suggestion
  if (serverState && serverState.port !== port) {
    // changing port, stop the old server
    serverState.disposable.dispose();
  }
```

Should be able to simplify this by combining the conditionals and getting rid of the branch that doesn't do anything.

Reply:
I just noticed the `return` keyword 🤦. Ignore this suggestion.

That said, I think just adding an `if` above this makes sense:

```ts
if (serverState?.port === port) {
  return;
}
```

Which would mean that any code after this would mean that `serverState.port` is different or uninitialized. I think then you can just remove the `if`. The whole thing would look like:

```ts
if (serverState?.port === port) {
  return;
}

serverState.disposable.dispose();
// ...
```

Reply:
c902746991f5d7fa339342b41226f05cb71ef4fb ✅

Reply:
and d4a65ee8cc6a0f82a3a83c43019c16ee8351bfab ... that took a moment to click for me